Modded Sand and Sandgun support

blahblahbal opened the pull request:
### What is the new feature?
Easier modded sand Sandgun ammo and general falling sand block support.

### Why should this be part of tModLoader?
Without this change, modded sand sandgun ammo requires a detour and some reflection (due to WorldGen.SpawnFallingBlockProjectile() being private). This change publicizes that method and adds 3 Sets (TileID, ProjectileID, ItemID) that allows for much easier sandgun ammo addition, as well as falling sand block support.

### Are there alternative designs?
Not that I'm aware of (unless you count using a detour and reflection).

### Sample usage for the new feature
See the ExampleMod updates section for a sample (I've included sand tile and item, falling sand projectile, and sandgun projectile in the updates I've done to ExampleMod).

### ExampleMod updates
I updated ExampleMod to add the previously mentioned files; these files should be pretty self-explanatory.

blahblahbal committed:
Modded Sand

- Modded Sand now works with the vanilla Sandgun via 4 Sets (two in ItemID and two in TileID)
- TileID.Sets.FallingBlockProjectileID and TileID.Sets.FallingBlockProjectileDamage should be set in your sand ModTile's SetStaticDefaults
- ItemID.Sets.SandgunAmmoToProjectile and ItemID.Sets.SandgunAmmoDamage should be set in your sand block item's SetStaticDefaults
- Also added ExampleSandBlock, ExampleFallingSandProjectile, ExampleSandSandgunProjectile, and ExampleSand (the tile) to ExampleMod

Rijam commented:
Make sure you follow the tModLoader style guide https://github.com/tModLoader/tModLoader/wiki/tModLoader-Style-Guide (mainly the K&R Style of curly braces).

I'm no authority here, but I think it would be better to put the declarations for the ItemID.Sets and TileID.Sets in `ItemID.TML.cs` and `TileID.TML.cs` instead of the patch file to reduce on patches. Also, you could use the full name IDs instead of the numeric IDs for readability if you want (you already did that for your ItemID.Sets, but you could do it for the TileID.Sets, too).

blahblahbal commented:
Oh, I missed the bracket style thing. I'll fix and push again. I can also put the Sets in the TML files. Didn't realize that was an option haha. (Will also change the numeric IDs to their appropriate named IDs)

blahblahbal commented:
Okay so the diff is confusing me; it _appears_ like the Player.cs.patch file still has the old code for the "if (projToShoot == 42)" section, but I've changed it? Might just be github desktop messing with my mind but the old lines aren't highlighted in red.

blahblahbal committed:
Fixes to adhere to guidelines

steviegt6 commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -7441,6 +7441,27 @@
+ 			projToShoot = 117;
+ 
+ 		if (projToShoot == 42) {
+-			if (item.type == 370) {
+-				projToShoot = 65;
+-				Damage += 5;
+-			}
+-			else if (item.type == 408) {
+-				projToShoot = 68;
+-				Damage += 5;
+-			}
+-			else if (item.type == 1246) {
+-				projToShoot = 354;
+-				Damage += 5;
+-			}
++			projToShoot = ItemID.Sets.SandgunAmmoToProjectile[item.type];
++			Damage += ItemID.Sets.SandgunAmmoDamage[item.type];
```
It's preferred to use multi-line comments (`/* ... */`) to remove large swathes of code. This minimizes exposure of ReLogic code in patches and makes them more resilient to change.

blahblahbal replied:
Ah, alright. Couldn't find anything about this specifically in the guidelines page (might've missed it though). I'll change it to comment out the code instead of just deleting it.

steviegt6 commented on patches/tModLoader/Terraria/WorldGen.cs.patch:
```diff
@@ -2962,6 +2971,60 @@
+-			case 112:
+-				projType = 56;
+-				break;
+-			case 59:
+-				projType = 39;
+-				break;
+-			case 116:
+-				projType = 67;
+-				break;
+-			case 123:
+-				projType = 71;
+-				break;
+-			case 224:
+-				projType = 179;
+-				break;
+-			case 234:
+-				projType = 241;
+-				break;
+-			case 330:
+-				projType = 411;
+-				dmg = 0;
+-				break;
+-			case 331:
+-				projType = 412;
+-				dmg = 0;
+-				break;
+-			case 332:
+-				projType = 413;
+-				dmg = 0;
+-				break;
+-			case 333:
+-				projType = 414;
+-				dmg = 0;
+-				break;
+-			case 495:
+-				projType = 812;
+-				break;
+-			default:
+-				projType = 31;
+-				break;
```
See my previous comment about handling large deletions.

steviegt6 commented on patches/tModLoader/Terraria/ID/TileID.TML.cs:
```diff
@@ -16,6 +16,28 @@ partial class Sets
+		public static int[] FallingBlockProjectileID = Factory.CreateIntSet(-1,
+			Sand, ProjectileID.SandBallFalling,
+			Ebonsand, ProjectileID.EbonsandBallFalling,
+			TileID.Mud, ProjectileID.MudBall,
```
Any reason you qualify `Mud` here (and farther below).

blahblahbal replied:
It needs it cause there's a TileID.Sets.Mud as well (probably used for the jungle? I didn't actually check)

Rijam replied:
Mud Blocks were affected by gravity in 1.0.5, then it was reverted in 1.0.6 (same with Ash) https://terraria.wiki.gg/wiki/Mud_Block#History
What tomat is talking about is how you put `TileID.Mud` instead of just `Mud`.

blahblahbal replied:
Yeah, in TileID there's a `TileID.Sets.Mud`, which if I use just `Mud`, it gives an error, since `TileID.Sets.Mud` is a bool array and can't be used as an `int`. The exact error is "Argument 6: cannot convert from 'bool[]' to 'int'".

steviegt6 commented on patches/tModLoader/Terraria/ID/ItemID.TML.cs:
```diff
@@ -6,6 +6,17 @@ partial class ItemID
+		/// <summary>
+		/// The damage a specific sandgun ammo should increase by. Defaults to 0.
+		/// </summary>
+		public static int[] SandgunAmmoDamage = Factory.CreateIntSet(0, EbonsandBlock, 5, PearlsandBlock, 5, CrimsandBlock, 5);
```
Should this be titled `SandgunAmmoBonusDamage` instead? Something to communicate its additive properties.

blahblahbal replied:
Yeah that's probably better actually. I'll change it.

blahblahbal committed:
More fixes to adhere to guidelines

- Re-add vanilla code (but commented out)
- Rename ItemID.Sets.SandgunAmmoDamage to ItemID.Sets.SandgunAmmoBonusDamage

blahblahbal commented:
Just pushed my changes. If there's anything else that I missed or need to do please let me know.

steviegt6 commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -7441,6 +7441,27 @@
+ 
+ 		if (projToShoot == 42) {
++			/*
+-			if (item.type == 370) {
++			 if (item.type == 370) {
```
Remove the space.

blahblahbal replied:
Fixed. I'm not sure how I missed that one when I didn't miss the other space in the comment in WorldGen.cs.

blahblahbal committed:
Fix extraneous space

jjohnsnaill commented:
Is `SandgunAmmoToProjectile` necessary? `AmmoID.Sets.SpecificLauncherAmmoProjectileMatches` should be sufficient, and if a mod adds its own sand-shooting weapons, it could simply reference the Sandgun's dictionary.

JavidPack committed:
Lots of changes, docs, etc

JavidPack commented:
I've made a lot of changes. The original AI code seemed to be based on an older Terraria version, it had some rare edge cases where things behaved slightly different (regarding actuators I think). It also seemed felt weird to me that the tile and projectile code still required large sections of copied code, so I updated the sets and patches to support things more directly.

Please test my changes and let me know if the set names make sense. Let me know if the new approach seems reasonable or any other feedback.

blahblahbal commented:
Seems to all work fine from my tests. And the set names make sense to me as well. Didn't realize that the projectile AI code was outdated - I thought I copied it directly from vanilla 1.4.4.9. Unless you meant the old ExampleMod code? Anyway, I don't have any other feedback; your approach seems reasonable enough as well.

JavidPack committed:
Combine tandem sets into SandgunAmmoProjectileData

JavidPack committed:
Merge branch '1.4.4' into 1.4.4
